### PR TITLE
Drop support for forward compatibility

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Removes forward compatibility support from Firecracker binary.

## Reason

Forward compatibility refers to the ability of a Firecracker binary to create snapshots for older versions of Firecracker.
We have found this feature to not be useful, while it increases code and testing complexity.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by roypat** [comment]
I think we might also want to remove https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L537-L551 with this.

**Comment by bchalios** [comment]
> I think we might also want to remove
> 
> https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L537-L551
> with this.

Why remove it? I think this will be reworked once we implement the new snapshot versioning schema. Once we do that work, each Firecracker version will declare support for a single `MAJOR.MINOR` snapshot version which will be decoupled from Firecracker version, but in any case I think it'd be worth being able to print it.

File: src/api_server/src/lib.rs [hunk 1]
```diff
@@ -295,7 +294,6 @@ mod tests {
                 snapshot_type: SnapshotType::Diff,
                 snapshot_path: PathBuf::new(),
                 mem_file_path: PathBuf::new(),
-                version: None,
```
Comment (ShadowCurse) [inline src/api_server/src/lib.rs:298]: Shouldn't we also remove `ser_fn` functions like [this](https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/vmm/src/vstate/vcpu/aarch64.rs#L295) from versionize? They are only used for creation of old versions of snaps.

**Comment by bchalios** [comment]
> > > I think we might also want to remove
> > > https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L537-L551
> > > 
> > > with this.
> > 
> > 
> > Why remove it? I think this will be reworked once we implement the new snapshot versioning schema. Once we do that work, each Firecracker version will declare support for a single `MAJOR.MINOR` snapshot version which will be decoupled from Firecracker version, but in any case I think it'd be worth being able to print it.
> 
> Mh, to me this output always read as "we can create snapshots for these firecracker versions". I guess if it also means "we can restore from these" then its indeed not redundant.
> 
> On the other hand, it just prints every firecracker version since we introduced snapshotting, which is completely wrong since there has been multiple non-compatible changes to the snapshot format since then. Do we plan to introduce the new versioning scheme with this release? If yes we can keep it as is (it'll get fixed when the new scheme gets introduced, as you say), but if no I'd rather we remove it (or fix it for the current version scheme).

Did you maybe mean that we should remove this function: https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L513-L524?


Then I agree. `print_snapshot_data_format` does not print the supported versions, so I got confused.

File: src/api_server/src/lib.rs (hunk 1 above)
Comment (bchalios) [reply src/api_server/src/lib.rs:298]: You are right! I will grep to find a list of these and remove.

**Comment by roypat** [comment]
> > > > I think we might also want to remove
> > > > https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L537-L551
> > > > 
> > > > with this.
> > > 
> > > 
> > > Why remove it? I think this will be reworked once we implement the new snapshot versioning schema. Once we do that work, each Firecracker version will declare support for a single `MAJOR.MINOR` snapshot version which will be decoupled from Firecracker version, but in any case I think it'd be worth being able to print it.
> > 
> > 
> > Mh, to me this output always read as "we can create snapshots for these firecracker versions". I guess if it also means "we can restore from these" then its indeed not redundant.
> > On the other hand, it just prints every firecracker version since we introduced snapshotting, which is completely wrong since there has been multiple non-compatible changes to the snapshot format since then. Do we plan to introduce the new versioning scheme with this release? If yes we can keep it as is (it'll get fixed when the new scheme gets introduced, as you say), but if no I'd rather we remove it (or fix it for the current version scheme).
> 
> Did you maybe mean that we should remove this function:
> 
> https://github.com/firecracker-microvm/firecracker/blob/9b0e03a437c753d40c29647e63f35f4fdb249fce/src/firecracker/src/main.rs#L513-L524
> ?
> 
> Then I agree. `print_snapshot_data_format` does not print the supported versions, so I got confused.

Oh, yes, so sorry! That's the one I was looking at in my editor, seems I failed at cross-referencing with GitHub

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 2]
```diff
@@ -35,6 +41,34 @@ def _get_guest_drive_size(ssh_connection, guest_dev_name="/dev/vdb"):
...
+    snapshot = vm.snapshot_full()
+
+    # Fetch Firecracker binary for the latest version
+    fc_binary, _ = get_firecracker_binaries()
+    # Verify the output of `--describe-snapshot` command line parameter
+    cmd = [str(fc_binary)] + ["--describe-snapshot", str(snapshot.vmstate)]
+
+    code, stdout, stderr = run_cmd(cmd)
+    assert code == 0, stderr
+    assert stderr == ""
```
Comment (roypat) [inline tests/integration_tests/functional/test_snapshot_basic.py:68]: nit: `run_cmd` already does this check

**Review by roypat — COMMENTED** [review]
I think we can also do

```diff
diff --git a/src/snapshot/src/lib.rs b/src/snapshot/src/lib.rs
index 8b394c4a4..9dfea9a3d 100644
--- a/src/snapshot/src/lib.rs
+++ b/src/snapshot/src/lib.rs
@@ -74,8 +74,6 @@ struct SnapshotHdr {
 pub struct Snapshot {
     hdr: SnapshotHdr,
     version_map: VersionMap,
-    // Required for serialization.
-    target_version: u16,
 }
 
 // Parse a magic_id and return the format version.
@@ -93,11 +91,10 @@ fn build_magic_id(format_version: u16) -> u64 {
 
 impl Snapshot {
     /// Creates a new instance which can only be used to save a new snapshot.
-    pub fn new(version_map: VersionMap, target_version: u16) -> Snapshot {
+    pub fn new(version_map: VersionMap) -> Snapshot {
         Snapshot {
             version_map,
             hdr: SnapshotHdr::default(),
-            target_version,
         }
     }
 
@@ -194,7 +191,7 @@ impl Snapshot {
         O: Versionize + Debug,
     {
         self.hdr = SnapshotHdr {
-            data_version: self.target_version,
+            data_version: self.version_map.latest_version(),
         };
 
         let format_version_map = Self::format_version_map();
@@ -216,7 +213,7 @@ impl Snapshot {
 
         // Serialize the object using the state version map.
         object
-            .serialize(&mut writer, &self.version_map, self.target_version)
+            .serialize(&mut writer, &self.version_map, self.version_map.latest_version())
             .map_err(Error::Versionize)?;
         writer
             .flush()
diff --git a/src/vmm/src/persist.rs b/src/vmm/src/persist.rs
index 7002426b4..d502cc7f3 100644
--- a/src/vmm/src/persist.rs
+++ b/src/vmm/src/persist.rs
@@ -209,8 +209,6 @@ pub fn create_snapshot(
     params: &CreateSnapshotParams,
     version_map: VersionMap,
 ) -> Result<(), CreateSnapshotError> {
-    let snapshot_data_version = version_map.latest_version();
-
     let microvm_state = vmm
         .save_state(vm_info)
         .map_err(CreateSnapshotError::MicrovmState)?;
@@ -218,7 +216,6 @@ pub fn create_snapshot(
     snapshot_state_to_file(
         &microvm_state,
         &params.snapshot_path,
-        snapshot_data_version,
         version_map,
     )?;
 
@@ -230,7 +227,6 @@ pub fn create_snapshot(
 fn snapshot_state_to_file(
     microvm_state: &MicrovmState,
     snapshot_path: &Path,
-    snapshot_data_version: u16,
     version_map: VersionMap,
 ) -> Result<(), CreateSnapshotError> {
     use self::CreateSnapshotError::*;
@@ -241,7 +237,7 @@ fn snapshot_state_to_file(
         .open(snapshot_path)
         .map_err(|err| SnapshotBackingFile("open", err))?;
 
-    let mut snapshot = Snapshot::new(version_map, snapshot_data_version);
+    let mut snapshot = Snapshot::new(version_map);
     snapshot
         .save(&mut snapshot_file, microvm_state)
         .map_err(SerializeMicrovmState)?;
```
and then fix up the tests accordingly

**Comment by bchalios** [comment]
> I think we can also do
> 
> ```diff
> diff --git a/src/snapshot/src/lib.rs b/src/snapshot/src/lib.rs
> index 8b394c4a4..9dfea9a3d 100644
> --- a/src/snapshot/src/lib.rs
> +++ b/src/snapshot/src/lib.rs
> @@ -74,8 +74,6 @@ struct SnapshotHdr {
>  pub struct Snapshot {
>      hdr: SnapshotHdr,
>      version_map: VersionMap,
> -    // Required for serialization.
> -    target_version: u16,
>  }
>  
>  // Parse a magic_id and return the format version.
> @@ -93,11 +91,10 @@ fn build_magic_id(format_version: u16) -> u64 {
>  
>  impl Snapshot {
>      /// Creates a new instance which can only be used to save a new snapshot.
> -    pub fn new(version_map: VersionMap, target_version: u16) -> Snapshot {
> +    pub fn new(version_map: VersionMap) -> Snapshot {
>          Snapshot {
>              version_map,
>              hdr: SnapshotHdr::default(),
> -            target_version,
>          }
>      }
>  
> @@ -194,7 +191,7 @@ impl Snapshot {
>          O: Versionize + Debug,
>      {
>          self.hdr = SnapshotHdr {
> -            data_version: self.target_version,
> +            data_version: self.version_map.latest_version(),
>          };
>  
>          let format_version_map = Self::format_version_map();
> @@ -216,7 +213,7 @@ impl Snapshot {
>  
>          // Serialize the object using the state version map.
>          object
> -            .serialize(&mut writer, &self.version_map, self.target_version)
> +            .serialize(&mut writer, &self.version_map, self.version_map.latest_version())
>              .map_err(Error::Versionize)?;
>          writer
>              .flush()
> diff --git a/src/vmm/src/persist.rs b/src/vmm/src/persist.rs
> index 7002426b4..d502cc7f3 100644
> --- a/src/vmm/src/persist.rs
> +++ b/src/vmm/src/persist.rs
> @@ -209,8 +209,6 @@ pub fn create_snapshot(
>      params: &CreateSnapshotParams,
>      version_map: VersionMap,
>  ) -> Result<(), CreateSnapshotError> {
> -    let snapshot_data_version = version_map.latest_version();
> -
>      let microvm_state = vmm
>          .save_state(vm_info)
>          .map_err(CreateSnapshotError::MicrovmState)?;
> @@ -218,7 +216,6 @@ pub fn create_snapshot(
>      snapshot_state_to_file(
>          &microvm_state,
>          &params.snapshot_path,
> -        snapshot_data_version,
>          version_map,
>      )?;
>  
> @@ -230,7 +227,6 @@ pub fn create_snapshot(
>  fn snapshot_state_to_file(
>      microvm_state: &MicrovmState,
>      snapshot_path: &Path,
> -    snapshot_data_version: u16,
>      version_map: VersionMap,
>  ) -> Result<(), CreateSnapshotError> {
>      use self::CreateSnapshotError::*;
> @@ -241,7 +237,7 @@ fn snapshot_state_to_file(
>          .open(snapshot_path)
>          .map_err(|err| SnapshotBackingFile("open", err))?;
>  
> -    let mut snapshot = Snapshot::new(version_map, snapshot_data_version);
> +    let mut snapshot = Snapshot::new(version_map);
>      snapshot
>          .save(&mut snapshot_file, microvm_state)
>          .map_err(SerializeMicrovmState)?;
> ```
> 
> and then fix up the tests accordingly

Would you mind if we did this when we redefine the Snapshot structure, with the new versioning schema? At the moment, nothing actually uses `target_version` in the `Snapshot` type. The code that calls into this, just uses the latest version.

The reason why I want to avoid doing this now, is because this code is used by the snapshot_editor tool, to edit snapshots. It might edit older snapshots, so the `target_version` is needed.

**Review by roypat — APPROVED** [review]

File: src/api_server/src/lib.rs (hunk 1 above)
Comment (bchalios) [reply src/api_server/src/lib.rs:298]: done!

File: src/vmm/src/device_manager/mmio.rs [hunk 3]
```diff
@@ -294,7 +294,7 @@ impl MMIODeviceManager {
         &self.id_to_dev_info
     }
 
-    #[cfg(target_arch = "x86_64")]
+    #[cfg(all(target_arch = "x86_64", test))]
```
[the remark this answers is not in the document]
Comment (bchalios) [reply src/vmm/src/device_manager/mmio.rs]: moved it in the `test` module, there was already another method implemented there just for tests.

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 2 above)
Comment (bchalios) [reply tests/integration_tests/functional/test_snapshot_basic.py]: Thanks, didn't know that. Removed

**Review by ShadowCurse — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,9 @@
   Refactored error propagation to avoid logging and printing an error on
   exits with a zero exit code. Now, on successful exit
   "Firecracker exited successfully" is logged.
+- [#4194](https://github.com/firecracker-microvm/firecracker/pull/4194):
+  Removed support for creating Firecracker snapshots targetting older versions
```
[the remark this answers is not in the document]
Comment (bchalios) [reply CHANGELOG.md]: Done. Thanks :)

**Review by roypat — APPROVED** [review]

**Review by sudanl0 — APPROVED** [review]

**Pull request merged**